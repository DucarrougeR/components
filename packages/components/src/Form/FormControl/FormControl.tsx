import * as CSS from 'csstype'
import React from 'react'
import { ResponsiveValue } from 'styled-system'
import { BoxProps } from '../../Layout/Box'
import { Flex } from '../../Layout/Flex'

type ResponsiveAlignItemsValue = ResponsiveValue<CSS.AlignItemsProperty>
type ResponsiveFlexDirectionValue = ResponsiveValue<CSS.FlexDirectionProperty>
type ResponsiveJustifyContentValue = ResponsiveValue<CSS.JustifyContentProperty>

type FormControlDirections = 'left' | 'right' | 'top' | 'bottom'

export interface FormControlProps extends Omit<BoxProps<HTMLDivElement>, 'as'> {
  alignLabel?: FormControlDirections
}

export const FormControl: React.FC<FormControlProps> = ({
  alignLabel,
  ...props
}) => {
  interface FlexAlignment {
    alignItems?: ResponsiveAlignItemsValue
    flexDirection?: ResponsiveFlexDirectionValue
    justifyContent?: ResponsiveJustifyContentValue
  }
  const flexAlignment: FlexAlignment = {}
  const setFlexAlignment = () => {
    switch (alignLabel) {
      case 'left':
        flexAlignment.flexDirection = 'row'
        flexAlignment.alignItems = 'baseline'
        break
      case 'right':
        flexAlignment.flexDirection = 'row-reverse'
        flexAlignment.justifyContent = 'flex-end'
        flexAlignment.alignItems = 'baseline'
        break
      case 'bottom':
        flexAlignment.flexDirection = 'column-reverse'
        flexAlignment.justifyContent = 'flex-end'
        break
      case 'top':
      default:
        flexAlignment.flexDirection = 'column'
        break
    }
    return flexAlignment
  }
  return (
    <Flex {...setFlexAlignment()} {...props}>
      {props.children}
    </Flex>
  )
}
