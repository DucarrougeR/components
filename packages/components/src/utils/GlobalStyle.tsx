import { css, createGlobalStyle } from 'styled-components'

/*
 * @TODO - Explore local loading of font files.
 *

import openSansSemi from '../fonts/open-sans-600.woff2'
import openSansRegular from '../fonts/open-sans-regular.woff2'

const fontFaces = css`
  @font-face {
    font-family: 'Open Sans';
    font-style: normal;
    font-weight: 400;
    src: local('Open Sans Regular'), local('OpenSans-Regular'),
      url(${openSansRegular}) format('woff');
  }

  @font-face {
    font-family: 'Open Sans';
    font-style: normal;
    font-weight: 600;
    src: local('Open Sans SemiBold'), local('OpenSans-SemiBold'),
      url(${openSansSemi}) format('woff');
  }
`
*/

const fonts = css`
  @import url('https://fonts.googleapis.com/css?family=Open+Sans:400,600&display=swap');

  body,
  button,
  input,
  textarea,
  select {
    font-family: 'Open Sans', sans-serif;
  }
`

const reset = css`
  * {
    padding: 0;
    box-sizing: border-box;
  }

  ul,
  li,
  p,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 0;
  }

  a {
    text-decoration: none;
  }

  html,
  body {
    width: 100%;
    height: 100%;
    margin: 0;
  }

  body {
    background-color: ${props => props.theme.colors.palette.white};
  }
`

export const GlobalStyle = createGlobalStyle`
  ${fonts}
  ${reset}
`
