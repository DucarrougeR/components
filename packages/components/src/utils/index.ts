export * from './GlobalStyle'
export * from './useToggle'
