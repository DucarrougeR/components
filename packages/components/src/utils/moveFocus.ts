/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

export const getTabStops = (ref: HTMLElement): HTMLElement[] =>
  Array.from(
    ref.querySelectorAll(
      'a,button:not(:disabled),[tabindex="0"],[tabindex="-1"]'
    )
  )

export const moveFocus = (
  direction: number,
  initial: number,
  element?: HTMLElement | null
) => {
  if (element) {
    const tabStops = getTabStops(element)

    if (tabStops.length > 0) {
      let toFocus = tabStops.slice(initial)[0]
      if (
        document.activeElement &&
        tabStops.includes(document.activeElement as HTMLElement)
      ) {
        const next =
          tabStops.findIndex((f) => f === document.activeElement) + direction

        if (next === tabStops.length) return null
        if (!tabStops[next]) return null
        toFocus = tabStops[next]
      }
      toFocus.focus()
      return toFocus
    }
  }
  return null
}
