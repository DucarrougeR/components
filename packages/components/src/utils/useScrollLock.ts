import { RefObject, useEffect } from 'react'

export function useScrollLock(
  disabled?: boolean,
  allowScrollWithin?: RefObject<HTMLElement | null>
) {
  useEffect(() => {
    let scrollTop = window.scrollY
    let scrollTarget: EventTarget | HTMLElement | null = document

    function stopScroll(e: Event) {
      if (e.target !== null && e.target !== scrollTarget) {
        scrollTarget = e.target
        scrollTop =
          scrollTarget instanceof Element
            ? scrollTarget.scrollTop
            : window.scrollY
      }

      if (
        scrollTarget instanceof Element &&
        !(
          allowScrollWithin &&
          allowScrollWithin.current &&
          allowScrollWithin.current.contains(scrollTarget)
        )
      ) {
        scrollTarget.scrollTop = scrollTop
      } else if (scrollTarget === document) {
        window.scrollTo({ top: scrollTop })
      }
    }

    if (disabled) {
      window.removeEventListener('scroll', stopScroll, true)
    } else {
      window.addEventListener('scroll', stopScroll, true)
    }

    return () => {
      window.removeEventListener('scroll', stopScroll, true)
    }
  }, [disabled, allowScrollWithin])
}
