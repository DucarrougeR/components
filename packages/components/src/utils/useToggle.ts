import { useState } from 'react'

export interface UseToggleReturn {
  value: boolean
  setOn: () => void
  setOff: () => void
  toggle: () => void
}

export function useToggle(initialValue = false): UseToggleReturn {
  const [value, setValue] = useState(initialValue)
  const setOn = () => setValue(true)
  const setOff = () => setValue(false)
  const toggle = () => setValue(!value)
  return { value, setOn, setOff, toggle }
}
