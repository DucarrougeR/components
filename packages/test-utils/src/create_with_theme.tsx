import { theme } from '@looker/design-tokens'
import { render } from '@testing-library/react'
import { mount, shallow } from 'enzyme'
import 'jest-styled-components'
import * as React from 'react'
import { create } from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'

export const withThemeProvider = (Component: React.ReactElement<any>) => (
  <ThemeProvider theme={theme}>{Component}</ThemeProvider>
)

export const createWithTheme = (Component: React.ReactElement<any>) =>
  create(withThemeProvider(Component))

export const mountWithTheme = (Component: React.ReactElement<any>) =>
  mount(withThemeProvider(Component))

export const shallowWithTheme = (Component: React.ReactElement<any>) =>
  shallow(withThemeProvider(Component))

export const renderWithTheme = (Component: React.ReactElement<any>) =>
  render(withThemeProvider(Component))
