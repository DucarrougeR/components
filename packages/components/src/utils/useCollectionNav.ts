/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { FocusEvent, KeyboardEvent, useState } from 'react'
import { getNextFocusTarget } from './getNextFocusTarget'
import { getTabStops, moveFocus } from './moveFocus'
import { useWrapEvent } from './useWrapEvent'

const isMovingUp = (e: FocusEvent, element: HTMLElement) => {
  const previousFocus = getNextFocusTarget(e)
  const relationship =
    previousFocus && element.compareDocumentPosition(previousFocus as Node)
  return relationship === Node.DOCUMENT_POSITION_FOLLOWING
}

export interface UseCollectionNavProps<E extends HTMLElement> {
  element: E | null
  onKeyDown?: (e: KeyboardEvent<E>) => void
  onFocus?: (e: FocusEvent<E>) => void
  onBlur?: (e: FocusEvent<E>) => void
}

export const useCollectionNav = <E extends HTMLElement = HTMLElement>({
  element,
  onBlur,
  onFocus,
  onKeyDown,
}: UseCollectionNavProps<E>) => {
  const [focusedItem, setFocusedItem] = useState<HTMLElement | null>(null)
  const [focusInside, setFocusInside] = useState(false)

  const handleArrowKey = (
    e: KeyboardEvent<E>,
    direction: number,
    initial: number
  ) => {
    const newFocusedItem = moveFocus(direction, initial, element)
    if (newFocusedItem) {
      e.preventDefault()
      setFocusedItem(newFocusedItem)
    }
  }

  const handleKeyDown = (e: KeyboardEvent<E>) => {
    switch (e.key) {
      case 'ArrowUp':
        e.preventDefault()
        handleArrowKey(e, -1, -1)
        break
      case 'ArrowDown':
        e.preventDefault()
        handleArrowKey(e, 1, 0)
        break
    }
  }

  const handleFocus = (e: FocusEvent<E>) => {
    setFocusInside(true)
    if (e.target === element) {
      if (focusedItem && element.contains(focusedItem)) {
        focusedItem.focus()
      } else {
        const tabStops = getTabStops(element)
        const toFocus = isMovingUp(e, element)
          ? tabStops[tabStops.length - 1]
          : tabStops[0]
        toFocus.focus()
      }
    }
  }

  const handleBlur = () => {
    setFocusInside(false)
  }

  return {
    onBlur: useWrapEvent(handleBlur, onBlur),
    onFocus: useWrapEvent(handleFocus, onFocus),
    onKeyDown: useWrapEvent(handleKeyDown, onKeyDown),
    tabIndex: focusInside ? undefined : 0,
  }
}
