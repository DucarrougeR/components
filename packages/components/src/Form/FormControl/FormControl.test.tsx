import 'jest-styled-components'
import * as React from 'react'
import { createWithTheme } from '../../../../test/utils/create_with_theme'
import { FormControl } from './FormControl'

test('A FormControl', () => {
  const component = createWithTheme(<FormControl />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
