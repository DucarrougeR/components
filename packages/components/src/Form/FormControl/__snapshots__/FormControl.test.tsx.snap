// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A FormControl 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
/>
`;
